### PR TITLE
P2P signing on direct send

### PR DESCRIPTION
**Pull request by iulianpascalau**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- the direct send messages are not signed. This is not a security concern since the libp2p lib checks the keys anyway at handshake time. To ensure behavioral consistency, we will add signature to the direct send messages, anyway.
  
## Proposed Changes
- added direct send message signing to ensure behavior consistency in the p2p network messenger
- added exception in consensus interceptor to not allow messages without signature

## Testing procedure
- standard testing procedure
- half network run

### CODE REVIEW

File: p2p/libp2p/directSender.go [hunk 1]
```diff
@@ -237,15 +252,41 @@ func (ds *directSender) getOrCreateStream(conn network.Conn) (network.Stream, er
...
 	mes.Topic = &topic
 	mes.From = []byte(conn.LocalPeer())
 	mes.Seqno = seqno
 
-	return &mes
+	buff, err := mes.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (ssd04) [inline p2p/libp2p/directSender.go:267]: in libp2p there is the `SignPrefix` that's being added, should we think about it here, while we construct the message?

File: p2p/libp2p/directSender_test.go [hunk 2]
```diff
@@ -67,43 +67,73 @@ func createLibP2PCredentialsDirectSender() (peer.ID, libp2pCrypto.PrivKey) {
...
-// ------- NewDirectSender
-
-func TestNewDirectSender_NilContextShouldErr(t *testing.T) {
-	hs := &mock.ConnectableHostStub{}
-
-	var ctx context.Context = nil
-	ds, err := libp2p.NewDirectSender(ctx, hs, func(msg *pubsub.Message, fromConnectedPeer core.PeerID) error {
-		return nil
+func TestNewDirectSender(t *testing.T) {
+	t.Run("nil context", func(t *testing.T) {
```
Comment (ssd04) [inline p2p/libp2p/directSender_test.go:71]: t.Parallel() for main test?

File: p2p/libp2p/directSender.go (hunk 1 above)
Comment (iulianpascalau) [reply p2p/libp2p/directSender.go:267]: added

File: p2p/libp2p/directSender_test.go [hunk 3]
```diff
@@ -564,3 +649,34 @@ func TestDirectSender_ProcessReceivedDirectMessageFromMismatchesFromConnectedPee
...
 }
+
+func TestDirectSender_ProcessReceivedDirectMessageSignatureFails(t *testing.T) {
+	verifyCalled := false
+	expectedErr := fmt.Errorf("expected error")
+	ds, _ := libp2p.NewDirectSender(
+		context.Background(),
+		generateHostStub(),
+		blankMessageHandler,
+		&mock.P2PSignerStub{
```
Comment (AdoAdoAdo) [inline p2p/libp2p/directSender_test.go:660]: do we have a test with real signer?

File: p2p/libp2p/netMessenger_test.go [hunk 4]
```diff
@@ -1076,10 +1085,90 @@ func generateConnWithRemotePeer(pid core.PeerID) network.Conn {
...
+	wg := &sync.WaitGroup{}
+	chanDone := make(chan bool)
+	wg.Add(1)
+
+	go func() {
+		wg.Wait()
+		chanDone <- true
+	}()
+
+	prepareMessengerForMatchDataReceive(messenger2, msg, wg)
```
Comment (AdoAdoAdo) [inline p2p/libp2p/netMessenger_test.go:1128]: can you make it clear in the test that real sig verification is done?

**Review by sstanculeanu — COMMENTED** [review]
no other comments

File: p2p/libp2p/directSender.go [hunk 5]
```diff
@@ -237,15 +252,41 @@ func (ds *directSender) getOrCreateStream(conn network.Conn) (network.Stream, er
...
+	mes.Signature, err = ds.signer.Sign(buff)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply p2p/libp2p/directSender.go:268]: analyzed & fixed

File: p2p/libp2p/directSender_test.go (hunk 3 above)
Comment (iulianpascalau) [reply p2p/libp2p/directSender_test.go:660]: yes, in netMessenger as a semi-integration test
TestLibp2pMessenger_SendDirectWithRealMessengersShouldWork
TestLibp2pMessenger_SendDirectWithRealMessengersWithoutSignatureShouldWork

File: p2p/libp2p/directSender_test.go (hunk 2 above)
Comment (iulianpascalau) [reply p2p/libp2p/directSender_test.go:71]: done in all tests

File: p2p/libp2p/netMessenger_test.go (hunk 4 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger_test.go:1128]: added signature size checks on the refactored & new test

**Review by ssd04 — DISMISSED** [review]

File: p2p/libp2p/directSender.go [hunk 6]
```diff
@@ -27,6 +27,8 @@ var _ p2p.DirectSender = (*directSender)(nil)
 
 const timeSeenMessages = time.Second * 120
 const maxMutexes = 10000
+const signPrefix = "en-directsend:"
```
Comment (AdoAdoAdo) [inline p2p/libp2p/directSender.go:30]: maybe use the same prefix as in libp2p so that we can treat broadcast and direct send messages similarly.

Comment (iulianpascalau) [reply p2p/libp2p/directSender.go]: fixed, reused pubsub.SignPrefix

**Review by sstanculeanu — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Pull request merged**